lib/Common/Memory/LargeHeapBlock.cpp

```diff
@@ -970,8 +975,46 @@ LargeHeapBlock::ScanNewImplicitRoots(Recycler * recycler)
 }
 
 #if ENABLE_CONCURRENT_GC
+bool LargeHeapBlock::CheckDirtyOnePage(char* page, RescanFlags flags, bool isWriteBarrier)
+{
+#ifdef RECYCLER_WRITE_BARRIER
+    // TODO: SWB, use special page allocator for large block with write barrier?
+    if (CONFIG_FLAG(WriteBarrierTest))
+    {
+        Assert(isWriteBarrier);
+    }
+    if (isWriteBarrier)
+    {
+        return (RecyclerWriteBarrierManager::GetWriteBarrier(page) & DIRTYBIT) == DIRTYBIT;
+    }
+#endif
+
+#if ENABLE_WRITE_WATCH
+    if (!CONFIG_FLAG(ForceSoftwareWriteBarrier))
+    {
+        ULONG_PTR count = 1;
+        DWORD pageSize = AutoSystemInfo::PageSize;
+        DWORD const writeWatchFlags = (flags & RescanFlags_ResetWriteWatch ? WRITE_WATCH_FLAG_RESET : 0);
+        void * written = nullptr;
+        UINT ret = GetWriteWatch(writeWatchFlags, page, AutoSystemInfo::PageSize, &written, &count, &pageSize);
+        bool isDirty = (ret != 0) || (count == 1);
+        return isDirty;
+    }
+    else
+    {
+        Assert(UNREACHED);
```

Comment:
in case we disabled hardware write barrier, but we are still checking hardware write barrier on the page. yes it's better to be a fail fast, let me change it


---
In reply to: [88135185](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88135185) [](ancestors = 88135185)